Framework/DataHandling/src/ISISJournal.cpp

```diff
@@ -0,0 +1,278 @@
+    auto const childElement = element->getChildElement(filterKvp.first);
+    if (getTextValue(childElement) != filterKvp.second)
+      return false;
+  }
+  return true;
+}
+
+/* Extract a list of named "tag" values for the given element. Gets the text
+ * values of the child elements with the given names and returns a map of the
+ * tag name to the value. Also always adds the element name to the list so there
+ * is always a results for every run.
+ */
+ISISJournalData getTagsForNode(Element *element, ISISJournalTags const &tags) {
+  auto result = ISISJournalData{};
+  // Add the run name
+  result["name"] = element->getAttribute("name");
+  // Add any tags in the tags list
+  for (auto &tag : tags)
+    result[tag] = getTextValue(element->getChildElement(tag));
+  return result;
+}
+
+/* Extract a list of "tag" values for all child nodes in the given "root" or
+ * parent element. Here, a "tag" is a name for a child element in the element
+ * we're checking i.e. a grandchild of the root node.
+ */
+std::vector<ISISJournalData>
+getTagsForAllNodes(Element *root, ISISJournalTags const &tags,
+                   ISISJournalFilters const &filters) {
+  auto results = std::vector<ISISJournalData>{};
+
+  auto nodeIter = TreeWalker(root, NodeFilter::SHOW_ELEMENT);
+  for (auto node = nodeIter.nextNode(); node; node = nodeIter.nextSibling()) {
+    auto element = dynamic_cast<Element *>(node);
+    validateElement(element, NXENTRY_TAG);
+    if (matchesAllFilters(element, filters))
+      results.emplace_back(getTagsForNode(element, tags));
+  }
+
+  return results;
+}
+
+/* Get the text values for all direct child elements of a given parent
+ * element. The child elements should all have the given tag name - throws if
+ * not.
+ */
+std::vector<std::string> getTagValuesForChildElements(Element *root,
+                                                      const char *tag) {
+  auto results = std::vector<std::string>{};
+
+  auto nodeIter = TreeWalker(root, NodeFilter::SHOW_ELEMENT);
+  for (auto node = nodeIter.nextNode(); node; node = nodeIter.nextSibling()) {
+    auto element = dynamic_cast<Element *>(node);
+    validateElement(element, tag);
+    results.emplace_back(element->getAttribute("name"));
+  }
+
+  return results;
+}
+
+/* Convert a cycle filename to the cycle name or return an empty string if it
+ * doesn't match the required pattern
+ */
+std::string convertFilenameToCycleName(std::string const &filename) {
+  std::regex pattern("[0-9]+_[0-9]+");
+  std::smatch matches;
+  std::regex_search(filename, matches, pattern);
+
+  if (matches.size() == 1)
+    return matches[0];
+
+  return std::string();
+}
+
+/* Convert a list of cycle filenames to a list of cycle names e.g.
+ * journal_19_4.xml -> 19_4. Also exlcudes files from the list if they do not
+ * match the required pattern.
+ */
+std::vector<std::string>
+convertFilenamesToCycleNames(std::vector<std::string> const &filenames) {
+  std::vector<std::string> cycles;
```

Comment:
You'll want to do do
cycles.reserve(filenames.size())

As len(cycles) <= len(filenames)